anndata/_core/anndata.py

```diff
@@ -914,8 +910,14 @@ def var_names(self) -> pd.Index:
     def var_names(self, names: Sequence[str]):
         utils.warn_no_string_index(names)
         self._var.index = names
-        if not self._var.index.is_unique:
-            utils.warn_names_duplicates("var")
+        self._normalize_index(self._var.index, "var")
+
+    @staticmethod
+    def _normalize_index(idx: pd.Index, idx_name: str):
+        if isinstance(idx.name, (int, np.integer)):
+            idx.name = None
```

Comment:
@flying-sheep, I'm not sure I like this. This lets the expression `adata.var_names = foo` mutate the RHS.

Reply:
No, because I pass `self._var.index`/`self._obs.index`

Reply:
No, because I pass `self._var.index`/`self._obs.index`

Reply:
I think that still mutates the original, since no copy has been made. Here's an example:

```python
import pandas as pd
a = pd.DataFrame({"a": [1,2,3,4]})
b = a.copy()
b.index = a.index
b.index.name = "b" 
print(a.index.name)
"b"
```

Reply:
Oh, so no copy is created implicitly when setting a DataFrame’s index, whoops. Will fix.

Reply:
Fixed in #294